### PR TITLE
8282208: Reduce MachNode size

### PR DESCRIPTION
**Pull request by neliasso**
Hi,

This patch reduces the size of MachNode from 72  to 64 bytes in product builds, and from 120 to 104 bytes in debug builds. In debug builds the size of Node is reduced from 104 to 96 bytes.

The motivation for this patch is that I needed more bits for the _barrier_data field. (This will happen in another patch.) I noticed that the bool _remove field caused a lot of padding. It's only used by peephole and can as well be modeled as a Node::flag. This change together with a slight rearrangement of fields a lot of unnecessary padding can be eliminated. 

Fixes in this patch:

1) Change MachNode::_remove to a Node::flag

2) Changing MachNode::operands to a 16 bit unsigned (Allowing room for 16 bits _barrier_data later) 

3) Moving Node::_debug_idx below Node::_debug_orig saves 8 bytes in debug builds. This is achieved by having an even number of 4 bytes fields in a row, eliminating two 4 byte paddings.

Please review,
Nils Eliasson

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282208](https://bugs.openjdk.java.net/browse/JDK-8282208): Reduce MachNode size

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 88145780ff364efc585180ef532879d0450c4254
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Jie Fu](https://openjdk.java.net/census#jiefu) (@DamonFool - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7562/head:pull/7562` \
`$ git checkout pull/7562`

Update a local copy of the PR: \
`$ git checkout pull/7562` \
`$ git pull https://git.openjdk.java.net/jdk pull/7562/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7562`

View PR using the GUI difftool: \
`$ git pr show -t 7562`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7562.diff">https://git.openjdk.java.net/jdk/pull/7562.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by neliasso** [comment]
Thank you @vnkozlov !

I can add that I have listed the node layouts in the bug report for reference.

**Comment by DamonFool** [comment]
There are several failures in the github tests.
Did you test the failing cases with your patch?

Also, we'd better update the copyright year.
Thanks.

File: src/hotspot/share/opto/node.hpp [hunk 1]
```diff
@@ -786,6 +786,7 @@ class Node {
     Flag_is_expensive                = 1 << 14,
     Flag_is_predicated_vector        = 1 << 15,
     Flag_for_post_loop_opts_igvn     = 1 << 16,
+    Flag_is_removed_by_peephole      = 1 << 17,
     _last_flag                       = Flag_for_post_loop_opts_igvn
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/node.hpp:790]:
```suggestion
    Flag_is_removed_by_peephole      = 1 << 17,
    _last_flag                       = Flag_is_removed_by_peephole
```

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
The value of `Node::_last_flag` needs to be updated.

File: src/hotspot/share/opto/node.hpp (hunk 1 above)
Comment (neliasso) [reply src/hotspot/share/opto/node.hpp]: Good catch! Thanks!

**Comment by neliasso** [comment]
> There are several failures in the github tests. Did you test the failing cases with your patch?
> 
> Also, we'd better update the copyright year. Thanks.

I looked at the test failures, and they look unrelated. I've also run separate testing that didn't show these failures.

The github tests will run again now.

Updated copyright year.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by DamonFool — APPROVED** [review]
LGTM

**Pull request closed without merging**